Source/WebCore/platform/graphics/mac/controls/ControlMac.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if PLATFORM(MAC)
+
+#import "LengthBox.h"
+#import "PlatformControl.h"
+
+namespace WebCore {
+
+class ControlFactoryMac;
+class FloatSize;
+class GraphicsContext;
+class IntSize;
+
+class ControlMac : public PlatformControl {
+public:
+    ControlMac(ControlPart&, ControlFactoryMac&);
+
+protected:
+    void setFocusRingClipRect(const FloatRect& clipBounds) override;
+
+    void updateCellStates(const FloatRect&, const ControlStyle&) override;
+
+    void drawCellOrFocusRing(GraphicsContext&, const FloatRect&, const ControlStyle&, NSCell *, NSView *, bool drawCell = true);
+    void drawCellOrFocusRing(GraphicsContext&, const FloatRect&, float deviceScaleFactor, const ControlStyle&, NSCell *, NSView *, bool drawCell = true);
```

Comment:
These two functions have surprisingly different behavior; one can use an intermediate buffer. The other one will just draw a wrongly scaled thing in scaled contexts. Either give them different names, or make one of them private.

Reply:
Fixed. I renamed the one that creates the ImageBuffer to `ControlMac::draw()`. I kept the second one as is but I removed it as a class method and I made it a static function.